### PR TITLE
Validate and filter parameter and return value documentation

### PR DESCRIPTION
**Pull request by d-ronnqvist**
Bug/issue #, if applicable: rdar://118739612

## Summary

This adds validation and filtering of parameter and return value documentation based on a symbol's function signature in each language representation.

This results in 3 new behaviors:
- A parameter or return value that's not applicable to a specific language representation of a symbol isn't displayed in that language's variant of the page.
- Objective-C error parameters or return values that are not documented for symbols defined in other languages are synthesized if any other parameters are documented.
- User facing diagnostics with fixits are emitted for parameters that are misspelled, not found, or repeated and for return values for symbols that return void in all language representations.

### Filtering parameters

If a symbol has different parameters and return values in different languages, then each language's version of that symbol's page will only display the parameters that's applicable to that language. 

This is for example relevant for API with errors in Objective-C that bridge to API that `throws` in Swift;

```objc
/// Does something in Objective-C
/// - Parameters:
///   - someValue: Some value.
///   - error: On input, a pointer to an error object. If an error occurs, this pointer is set to an actual error object containing the error information.
/// - Returns: `YES` if doing something was successful, or `NO` if an error occurred.
- (BOOL)doSomethingWith:(NSInteger)someValue
                  error:(NSError **)error;
```

This method become `func doSomething(with someValue: Int) throws` in Swift which returns `Void` and doesn't have an `error` parameter, so displaying documentation for the error parameter and the return value wouldn't be helpful for someone reading the Swift version of that method's documentation. 

Before this PR we displayed all documented parameters in all language variants of a symbol's page. With these changes, the Swift page won't display the "error" parameter documentation or the return value documentation.

### Adding Objective-C error documentation

The above works well for functions defined in Objective-C that document parameters and return values that aren't available in Swift, but it doesn't work in the other direction;

```swift
/// Does something in Swift.
/// - Parameter someValue: Some value.
/// - Throws: Some error if something does wrong
@objc public func doSomething(with someValue: Int) throws { }
```

This becomes `- (BOOL)doSomethingWith:(NSInteger)someValue error:(NSError **)error;` in Objective-C but the Swift documentation comment doesn't have any documentation for the Objective-C "error" parameter or the Objective-C specific return value.

Before this PR we only displayed the developer documented parameters. With these changes, the Objective-C page will display a synthesized "error" parameter documentation with the generic description:

> **error**
> On input, a pointer to an error object. If an error occurs, this pointer is set to an actual error object containing the error information. 

and a synthesized return value documentation with the generic description:
> **Return Value**
>`YES` if successful, or `NO` if an error occurred.

If the throwing Swift function was defined with a non-null return value that bridges to a nullable return value in Objective-C;

```swift
/// Returns something from Swift.
/// - Returns: Some string.
/// - Throws: Some error if something does wrong
@objc public func returnSomething() throws -> String { "" }
```

Then the Objective-C return value documentation is appended with a generic description of this automatic bridging behavior:

> **Return Value**
> Some string. If an error occurs, this method returns `nil` and assigns an appropriate error object to the `error` parameter.

### Parameter and return value diagnostics 

If the symbol documents at least one parameter or its return value, then DocC will validate that symbol's parameters section and return value section. Specifically, DocC will warn in 5 cases:
1. A parameter is documented more than once
2. A parameter that doesn't exist in the function signature is documented
3. A parameter from the function signature isn't documented
4. A parameter is documented using its argument label instead of its parameter name
5. A symbol that returns `Void` has documented a return value

The Swift function below shows minimal examples of each of these cases:

```swift
/// - Parameters:
///   - first: Something
///   - first: Something else (1)
///   - with: Something (4)
///   - third: Something (2)
///  (3)
/// - Returns: Something (5)
func doSomething(with first: Int, and second: Int) { }
```

## Dependencies

https://github.com/apple/swift-docc-symbolkit/pull/64

## Testing

- In a multi-language project, add:
  - Objective-C API with errors with documentation for all parameters and return values
  - Swift `@objc` API that `throws` with documentation for the Swift parameters.

- Build documentation for the project and pass `--enable-experimental-parameters-and-returns-validation`.

- View both the Swift and Objective-C versions of the documentation pages for both the API defined in Objective-C and the API defined in Swift. 
  - Each language variant of the documentation pages should only display the parameters and return values applicable to that language. 

- Add parameter documentation for parameters that doesn't exist to the Swift and Objective-C definitions and build documentation again. 

- View both the Swift and Objective-C versions of the documentation pages for both the API defined in Objective-C and the API defined in Swift. 
  - The documentation for the parameters that doesn't exist shouldn't be displayed on either languages's variant of the documentation page.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- [x] Updated documentation if necessary

### CODE REVIEW

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+                variants[trait] = ParametersSection(parameters: [])
+                continue
+            }
+            
+            // Add a missing error parameter documentation if needed.
+            if trait == .objectiveC, Self.shouldAddObjectiveCErrorParameter(signatures, parameters) {
+                languageApplicableParameters["error"] = [Parameter(name: "error", contents: Self.objcErrorDescription)] // This parameter is synthesized and doesn't have a source range.
+            }
+            
+            // Ensure that the parameters are displayed in the order that they're declared.
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:119]: Why? Shouldn't authors have the flexibility to display parameters in any order?

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+            var sortedParameters: [Parameter] = []
+            sortedParameters.reserveCapacity(languageApplicableParameters.count)
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:121]: Maybe this is a common Swift idiom, but it seems inefficient to adjust the capacity after creating the array. Is there any Array initializer that allows you to create the array with a given capacity?

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+            for parameter in signature.parameters {
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:122]: I feel like you might be able to refactor this function and avoid these separate sorting loops for each trait. Above you iterate over the signature parameter also... If you could find a way to add each language's parameter in the loop above then the result array would already be sorted.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+        var argumentLabelsAndPossibleParameterNames: [String: Set<String>] = [:]
+        // Group the parameters by name to be able to diagnose parameters documented more than once.
+        let parametersByName = [String: [Parameter]](grouping: parameters, by: \.name)
+        
+        for (trait, signature) in signatures {
+            // Collect all language representations's parameter information from the function signatures.
+            let knownParameterNames = Set(signature.parameters.map(\.name))
+            allKnownParameterNames.formUnion(knownParameterNames)
+            for parameter in signature.parameters {
+                if let externalName = parameter.externalName {
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:101]: Is `externalName` the second label that Swift functions can use, example match vs. foundDocumentationExtension in:

```
private func nodeWithInitializedContent(reference: ResolvedTopicReference, match foundDocumentationExtension: DocumentationContext.SemanticResult<Article>?) -> DocumentationNode {
```

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+                    argumentLabelsAndPossibleParameterNames[externalName, default: []].insert(parameter.name)
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:102]: Assuming the answer is yes, then maybe we can think of a better name for `argumentLabelsAndPossibleParameterNames`? Something like `parameterNamesByExternalName` or something similar that indicates the keys are the external names.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+            possibleSolutions: nearMisses.map { candidate in
+                Solution(
+                    summary: "Replace \(parameter.name.singleQuoted) with \(candidate.singleQuoted)",
+                    replacements: parameter.nameRange.map { [Replacement(range: adjusted($0), replacement: candidate)] } ?? []
+                )
+            }
+        )
+    }
+    
+    private func externalParameterNameProblem(_ parameter: Parameter, knownParameterNamesForArgumentLabel: [String]) -> Problem {
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:367]: This diagnostic is especially confusing. Could you put an example of this (or ideally all of these diagnostics) in code comments to make it simpler for DocC developers to understand what's happening here?

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 5 above)
Comment (patshaughnessy) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:102]: Also, how is it possible for more than one parameter to have the same external name? That is, how could the size of a value in `argumentLabelsAndPossibleParameterNames` have a count > 1?

Maybe I'm misunderstanding the meaning of external name?

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+        // Diagnose parameters that are documented more than once.
+        for (name, parameters) in parametersByName where allKnownParameterNames.contains(name) && parameters.count > 1 {
+            let first = parameters.first! // Each group is guaranteed to be non-empty.
+            for parameter in parameters.dropFirst() {
+                diagnosticEngine.emit(duplicateParameterProblem(parameter, previous: first, symbolKind: symbolKind))
+            }
+        }
+        
+        // Diagnose missing parameters, based on the language representation with the most parameters.
+        if let parameterOrder = signatures.values.map(\.parameters).max(by: { $0.count < $1.count })?.map(\.name) {
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:149]: This line of code is especially confusing. Maybe add more explanation to the comment?

Also, the name `parameterOrder` doesn't help much to understand what this value is. If I'm reading this correctly, `parameterOrder` is an array of parameter names from the language/trait that has the most parameters.

A variable name such as `parameterNamesOrder` or `parameterNamesFromLongestSignature` or something similar would make it clear this is an array of names.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+            let parametersEndLocation = parameters.last?.range?.upperBound
+            
+            for parameterName in allKnownParameterNames.subtracting(["error"]) where parametersByName[parameterName] == nil {
+                // Look for the parameter that should come after the missing parameter to insert the placeholder documentation in the right location.
+                let parameterAfter = parameterOrder.drop(while: { $0 != parameterName }).dropFirst()
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:154]: Same here. Maybe rename `parameterAfter` to `parameterNameAfter` to make it clear this is a name, not a parameter.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+                identifier: "org.swift.docc.MissingParameterDocumentation",
+                summary: "Parameter \(name.singleQuoted) is missing documentation"
+            ),
+            possibleSolutions: solutions
+        )
+    }
+    
+    // MARK: Generated content
+    
+    private static let objcErrorDescription: [Markup] = [
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:456]: Is there a better place to save this generated content? Having text in the compiler source code seems like a maintenance problem waiting to happen. We might decide to change our style guide or possible an editor might want to make some changes to these text strings and now we need a new DocC compiler version/build.

And if we ever localized documentation someday, these English strings will be hard to track down and translate.

Does DocC have any text files or string files where we could save these sorts of literal text strings?

File: Sources/SwiftDocC/Model/Rendering/Variants/JSONPatchApplier.swift [hunk 10]
```diff
@@ -25,20 +25,18 @@ public struct JSONPatchApplier {
     /// - Returns: The JSON data with the patch applied.
     /// - Throws: This function throws an ``Error`` if the application was not successful.
     public func apply(_ patch: JSONPatch, to jsonData: Data) throws -> Data {
-        let json = try JSONDecoder().decode(JSON.self, from: jsonData)
+        var json = try JSONDecoder().decode(JSON.self, from: jsonData)
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/Rendering/Variants/JSONPatchApplier.swift:28]: Are you slipping in a performance fix here? Was there a bug earlier? Or is this just a coding style change?

File: Sources/SwiftDocC/Model/Semantics/Return.swift [hunk 11]
```diff
@@ -1,7 +1,7 @@
 /*
  This source file is part of the Swift.org open source project
 
- Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Copyright (c) 2021-2023 Apple Inc. and the Swift project authors
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/Semantics/Return.swift:4]: 2024 now

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 12]
```diff
@@ -0,0 +1,473 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2023 Apple Inc. and the Swift project authors
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:4]: It's 2024 now :)

...and same throughout this PR.

**Review by patshaughnessy — COMMENTED** [review]
Amazing work, and a great improvement for DocC!

I'm about halfway through reviewing this. I've read the code and have a number of questions for you. Most of my concerns are about understand your new algorithms... in some places it's a bit hard to follow. Some more code comments and different variable names might help a lot.

Next week I'll find some time to test this more carefully and review the test code.

File: Sources/SwiftDocC/Model/Semantics/Return.swift (hunk 11 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/Semantics/Return.swift:4]: It wasn't when I changed this file

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 12 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:4]: Like above, this change was made in 2023.

Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:4]: If I make additional changes to files I will update the years then.

File: Sources/SwiftDocC/Model/Rendering/Variants/JSONPatchApplier.swift (hunk 10 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/Rendering/Variants/JSONPatchApplier.swift:28]: Yes, this was a small performance fix for something that I stumbled upon. If you prefer I can extract it into a separate PR. I just find that I tend to forget if I don't fix things that I stumble upon.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 4 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:101]: No, external name is the name that's included in the symbol name. I [documented this in the SymbolKit PR.](https://github.com/apple/swift-docc-symbolkit/pull/64/files#diff-38139c158e2d8026197306fe59fe789160efbcc944f94f773b2e674372c53527R36-R53)

> In Swift which supports separate argument labels and parameter names, this function has one parameter where ``externalName`` is "with" and ``name`` is "someValue".
>
>```swift
>func doSomething(with someValue: Int) {}
>//               │    ╰─ name
>//               ╰─ externalName
>```

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 5 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:102]: > Also, how is it possible for more than one parameter to have the same external name? That is, how could the size of a value in `argumentLabelsAndPossibleParameterNames` have a count > 1?

For parameters with separate argument label (external name) and parameter name (name / internal name) then the external name is the first one in the declaration. 

For example, this poorly named function as two parameters with the external name "and":

```swift
func doSomething(with first: Int, and second: Int, and third: Int) {}
//               │    ╰─ name     │   ╰─ name      │   ╰─ name
//               ╰─ externalName  ╰─ externalName  ╰─ externalName
```

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 2 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:121]: No, there is no initializer that takes a specific capacity. Receiving capacity before adding a known number of elements is just a way to avoid multiple reallocations.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 5 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:102]: > Assuming the answer is yes, then maybe we can think of a better name for `argumentLabelsAndPossibleParameterNames`? Something like `parameterNamesByExternalName` or something similar that indicates the keys are the external names.

Sure, I can name this whatever is more readable. Do you prefer `yByX` over `xsAndYs` for a `[X: Y]` dictionary?

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 1 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:119]: My reasoning was that the parameters need to be passed in a specific order so I thought that it was best to ensure that they're also listed in that order.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 4 above)
Comment (patshaughnessy) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:101]: Yes - that's what I meant. Ok perfect thanks for the explanation.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 5 above)
Comment (patshaughnessy) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:102]: Personally I prefer `yByX`.... so `parameterNamesByExternalName `. I suppose I was confused by "argumentLabel" which I guess is the same as "externalName".

Comment (patshaughnessy) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:102]: I'm shocked Swift allows repeated external names/argument labels; thanks for the example above.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 7 above)
Comment (patshaughnessy) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:149]: Interesting - I prefer to try to name variables in a way that describes what they are/contain, and not what they are used for... The following lines of code will show the code reader what it is used for.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 13]
```diff
@@ -0,0 +1,473 @@
...
+            guard !languageApplicableParameters.isEmpty else { 
+                // Add an empty section so that this language doesn't fallback to another language's content.
+                variants[trait] = ParametersSection(parameters: [])
+                continue
+            }
+            
+            // Add a missing error parameter documentation if needed.
+            if trait == .objectiveC, Self.shouldAddObjectiveCErrorParameter(signatures, parameters) {
+                languageApplicableParameters["error"] = [Parameter(name: "error", contents: Self.objcErrorDescription)] // This parameter is synthesized and doesn't have a source range.
+            }
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:117]: I ran into a bug here while testing on an actual framework... I think you need to move the `shouldAddObjectiveCErrorParameter` check up ahead of the `guard` statement that tests `languageApplicableParameters.isEmpty`. I ran into this when I added invalid parameters to an Obj-C function that returned an error via an error parameter, but had no other parameters. What happens now is that the guard statement thinks there are no parameters, and so the error parameter is not added at all. Worse, DocC Render for some reason runs into trouble and fails to render the page at all, after a JavaScript error.

If you move the call to `shouldAddObjectiveCErrorParameter` up, the code will add the synthesized error parameter first, and then the `guard` statement will not set the parameter section to be empty.

If you're curious and want to try this yourself, here's the page I used as a test case: https://developer.apple.com/documentation/servicemanagement/smappservice/register()?language=objc

```suggestion
            // Remove documented parameters that don't apply to this language representation's function signature.
            var languageApplicableParameters = parametersByName.filter { knownParameterNames.contains($0.key) }
            
            // Add a missing error parameter documentation if needed.
            if trait == .objectiveC, Self.shouldAddObjectiveCErrorParameter(signatures, parameters) {
                languageApplicableParameters["error"] = [Parameter(name: "error", contents: Self.objcErrorDescription)] // This parameter is synthesized and doesn't have a source range.
            }

            guard !languageApplicableParameters.isEmpty else { 
                // Add an empty section so that this language doesn't fallback to another language's content.
                variants[trait] = ParametersSection(parameters: [])
                continue
            }
```

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 14]
```diff
@@ -0,0 +1,473 @@
...
+        else {
+            return false
+        }
+        
+        return true
+    }
+    
+    /// Checks if the validator should synthesize documentation for an Objective-C error parameter.
+    private static func shouldAddObjectiveCErrorParameter(_ signatures: Signatures, _ parameters: [Parameter]?) -> Bool {
+        parameters?.last?.name != "error" && hasSwiftThrowsObjectiveCErrorBridging(signatures)
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:257]: Is it possible for some other Obj-C parameter to be called "error"? If that is possible should we check for that before appending the new, synthesized error parameter? What would happen if we end up with two duplicate error parameters?

**Review by patshaughnessy — CHANGES_REQUESTED** [review]

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 8 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:154]: This is actually a `Parameter` value

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 1 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift]: I updated this documentation in https://github.com/apple/swift-docc/pull/776/commits/ef264671245be85daf65bb1ab7a3eb96a6d8a95b to clarify that it's the order that the developer needs to pass the parameters

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 4 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift]: I updated the documentation in https://github.com/apple/swift-docc/pull/776/commits/ef264671245be85daf65bb1ab7a3eb96a6d8a95b to clarify the difference between parameter names and external parameter names.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 5 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift]: I documented that parameters names are unique but external names aren't in https://github.com/apple/swift-docc/pull/776/commits/ef264671245be85daf65bb1ab7a3eb96a6d8a95b

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 13 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift]: Thanks. I will look at this next week.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 7 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift]: I documented the assumptions about why the language representation with the most parameters can determine the order in other language representations as well in https://github.com/apple/swift-docc/pull/776/commits/ef264671245be85daf65bb1ab7a3eb96a6d8a95b

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 13 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift]: I fixed the issue where DocC Render didn't display anything because the JSON patches got offset when the parameters section was completely empty. Then added a `testFunctionWithOnlyErrorParameter()` test that covers this example in  https://github.com/apple/swift-docc/pull/776/commits/844e530fcc705de9b88a5263831e35cf27997bf1

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@patshaughnessy I addressed all the feedback. Would you mind taking another look?

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 15]
```diff
@@ -0,0 +1,610 @@
...
+///
+/// To understand this file it helps to know the different between a parameter's "name" and its "external name".
+///
+///  - term name: The variable name that the parameter is bound in the implementation.
+///  - term external name: An optional "argument label" that appear instead of the parameter name at the call site.
+///
+/// ```swift
+/// func doSomething(with someValue: Int) {}
+/// //               │    ╰─ name
+/// //               ╰─ externalName (also known as "argument label")
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:25]: Thanks for this explanation! Love the ascii-art :)

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 16]
```diff
@@ -0,0 +1,610 @@
...
+            // Add a missing error parameter documentation if needed.
+            if trait == .objectiveC, Self.shouldAddObjectiveCErrorParameter(signatures, parameters) {
+                languageApplicableParametersByName["error"] = [Parameter(name: "error", contents: Self.objcErrorDescription)] // This parameter is synthesized and doesn't have a source range.
+            }
+            
+            // Ensure that the parameters are displayed in the order that they need to be passed.
+            var sortedParameters: [Parameter] = []
+            sortedParameters.reserveCapacity(languageApplicableParametersByName.count)
+            for parameter in signature.parameters {
+                if let foundParameter = languageApplicableParametersByName[parameter.name]?.first {
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:146]: Can we be sure `first` will return the proper parameter? Does it matter? Is the call to `(grouping: by:)` above guaranteed to retain the original order?

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 17]
```diff
@@ -0,0 +1,610 @@
...
+    }
+    
+    // MARK: Diagnostics
+    
+    /// Creates a new problem about return value documentation for a symbol that returns void.
+    ///
+    /// ## Example
+    ///
+    /// ```swift
+    /// /// - Paramaters:
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:372]: Typo/spelling - and throughout this appears in various places.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 18]
```diff
@@ -0,0 +1,610 @@
...
+    /// ///   - someValue: Some description of this parameter
+    /// /// - Returns: Description of what this function returns
+    /// ///   ^~~~~~~
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:375]: I love these examples of the error messages and how they will appear - thank you!

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift [hunk 19]
```diff
@@ -0,0 +1,610 @@
...
+    ///   - nextParameter: The next documented parameter for this symbol or `nil`, if the undocumented parameter is the last parameter for the symbol.
+    ///   - standalone: `true` if the existing documented parameters use the standalone syntax `- Parameter someValue:` or `false` if the existing documented parameters use the `- someValue` syntax within a `- Parameters:` list.
+    ///   - location: The end of the last parameter. Used as the diagnostic location when the undocumented parameter is the last parameter of the symbol.
+    /// - Returns: A new problem that suggests that the developer adds documentation for the parameter.
+    private func makeMissingParameterProblem(name: String, before nextParameter: Parameter?, standalone: Bool, lastParameterEndLocation: SourceLocation?) -> Problem {
+        let solutions: [Solution]
+        if let insertLocation = nextParameter?.range?.lowerBound ?? lastParameterEndLocation {
+            let extraWhitespace = "\n///" + String(repeating: " ", count: (nextParameter?.range?.lowerBound.column ?? 1 + (standalone ? 0 : 2) /* indent items in a parameter outline by 2 spaces */) - 1)
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:552]: Why include `///` here? Does that make the replacement ("Fix It" in the Xcode UI?) look better?

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 4 above)
Comment (patshaughnessy) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift]: thank you looks great!

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 9 above)
Comment (patshaughnessy) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:593]: Oh `embedInCode` sounds interesting. Looking forward to seeing that in use.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 16 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:146]: Yes, the order is [documented to be the same as the original order][1]. It mostly doesn't matter since documenting the same parameter results in a warning but it's nice that the behavior is stable across documentation builds even when there are warnings.

[1]: https://developer.apple.com/documentation/swift/dictionary/init(grouping:by:)

> The arrays in the “values” position of the new dictionary each contain at least one element, with the elements in the same order as the source sequence.

File: Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift (hunk 19 above)
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift:552]: Because the fix it inserts the placeholder text for the new parameter on its own line we need to include "///" after the newline so that the new line is part of the documentation comment.

For example, consider this declaration and documentation comment where the "first" parameter is undocumented and should be inserted before the "second".

```swift
/// - Paramaters:
///   - second: Some description of the first parameter
func someFunction(first: Int, second: Int) {}
```

Inserting only `- first: placeholder` without a newline would result in 

```swift
/// - Paramaters:
///   - first: placeholder- second: Some description of the first parameter
func someFunction(first: Int, second: Int) {}
```

Inserting only `- parameterName: placeholder\n` with a newline would result in 

```swift
/// - Paramaters:
///   - first: placeholder
- second: Some description of the first parameter
func someFunction(first: Int, second: Int) {}
```

But inserting `- parameterName: placeholder\n///   ` with a newline, a triple slash, and the same leading whitespace as the previous parameter had would result in 

```swift
/// - Paramaters:
///   - first: placeholder
///   - second: Some description of the first parameter
func someFunction(first: Int, second: Int) {}
```

**Comment by d-ronnqvist** [comment]
@swift-ci please test
